File: main/src/main/java/de/blinkt/openvpn/fragments/Settings_IP.java
/*
 * Copyright (c) 2012-2016 Arne Schwabe
 * Distributed under the GNU GPL v2 with additional terms. For full terms see the file doc/LICENSE.txt
 */

package de.blinkt.openvpn.fragments;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.EditTextPreference;
import android.preference.Preference;
import android.preference.Preference.OnPreferenceChangeListener;
import android.preference.PreferenceManager;
import android.preference.SwitchPreference;
import de.blinkt.openvpn.R;
import de.blinkt.openvpn.VpnProfile;

public class Settings_IP extends OpenVpnPreferencesFragment implements OnPreferenceChangeListener {
		private EditTextPreference mIPv4;
		private EditTextPreference mIPv6;
		private SwitchPreference mUsePull;
		private CheckBoxPreference mOverrideDNS;
		private EditTextPreference mSearchdomain;
		private EditTextPreference mDNS1;
		private EditTextPreference mDNS2;
		private CheckBoxPreference mNobind;

		@Override
		public void onCreate(Bundle savedInstanceState) {
			super.onCreate(savedInstanceState);
			
		
			// Make sure default values are applied.  In a real app, you would
			// want this in a shared function that is used to retrieve the
			// SharedPreferences wherever they are needed.
			PreferenceManager.setDefaultValues(getActivity(),
					R.xml.vpn_ipsettings, false);
			
			// Load the preferences from an XML resource
			addPreferencesFromResource(R.xml.vpn_ipsettings);
			mIPv4 = (EditTextPreference) findPreference("ipv4_address");
			mIPv6 = (EditTextPreference) findPreference("ipv6_address");
			mUsePull = (SwitchPreference) findPreference("usePull");
			mOverrideDNS = (CheckBoxPreference) findPreference("overrideDNS");
			mSearchdomain =(EditTextPreference) findPreference("searchdomain");
			mDNS1 = (EditTextPreference) findPreference("dns1");
			mDNS2 = (EditTextPreference) findPreference("dns2");
			mNobind = (CheckBoxPreference) findPreference("nobind");

			mIPv4.setOnPreferenceChangeListener(this);
			mIPv6.setOnPreferenceChangeListener(this);
			mDNS1.setOnPreferenceChangeListener(this);
			mDNS2.setOnPreferenceChangeListener(this);
			mUsePull.setOnPreferenceChangeListener(this);
			mOverrideDNS.setOnPreferenceChangeListener(this);
			mSearchdomain.setOnPreferenceChangeListener(this);
						
			loadSettings();
		}

		@Override
		protected void loadSettings() {

			mUsePull.setChecked(mProfile.mUsePull);
			mIPv4.setText(mProfile.mIPv4Address);
			mIPv6.setText(mProfile.mIPv6Address);
			mDNS1.setText(mProfile.mDNS1);
			mDNS2.setText(mProfile.mDNS2);
			mOverrideDNS.setChecked(mProfile.mOverrideDNS);
			mSearchdomain.setText(mProfile.mSearchDomain);
			mNobind.setChecked(mProfile.mNobind);
            if (mProfile.mAuthenticationType == VpnProfile.TYPE_STATICKEYS)
                mUsePull.setChecked(false);

			mUsePull.setEnabled(mProfile.mAuthenticationType != VpnProfile.TYPE_STATICKEYS);

			// Sets Summary
			onPreferenceChange(mIPv4, mIPv4.getText());
			onPreferenceChange(mIPv6, mIPv6.getText());
			onPreferenceChange(mDNS1, mDNS1.getText());
			onPreferenceChange(mDNS2, mDNS2.getText());
			onPreferenceChange(mSearchdomain, mSearchdomain.getText());

			setDNSState();
		}
		
		
		@Override
		protected void saveSettings() {
			mProfile.mUsePull = mUsePull.isChecked();
			mProfile.mIPv4Address = mIPv4.getText();
			mProfile.mIPv6Address = mIPv6.getText();
			mProfile.mDNS1 = mDNS1.getText();
			mProfile.mDNS2 = mDNS2.getText();
			mProfile.mOverrideDNS = mOverrideDNS.isChecked();
			mProfile.mSearchDomain = mSearchdomain.getText();
			mProfile.mNobind = mNobind.isChecked();
			
		}
		
		@Override
		public boolean onPreferenceChange(Preference preference,
				Object newValue) {
			if(preference==mIPv4 || preference == mIPv6 
					 || preference==mDNS1 || preference == mDNS2
					 || preference == mSearchdomain 
					)
			
				preference.setSummary((String)newValue);

			if(preference== mUsePull || preference == mOverrideDNS)
				if(preference==mOverrideDNS) { 
					// Set so the function gets the right value
					mOverrideDNS.setChecked((Boolean) newValue);
				}
				setDNSState();
			
			saveSettings();
			return true;
		}

		private void setDNSState() {
			boolean enabled;
			mOverrideDNS.setEnabled(mUsePull.isChecked());
			if(!mUsePull.isChecked())
				enabled =true;
			else
                enabled = mOverrideDNS.isChecked();

			mDNS1.setEnabled(enabled);
			mDNS2.setEnabled(enabled);
			mSearchdomain.setEnabled(enabled);
			

		}


	}

File: main/src/ui/java/de/blinkt/openvpn/fragments/Settings_IP.kt
/*
 * Copyright (c) 2012-2016 Arne Schwabe
 * Distributed under the GNU GPL v2 with additional terms. For full terms see the file doc/LICENSE.txt
 */
package de.blinkt.openvpn.fragments

import android.os.Bundle
import android.view.View
import androidx.preference.*
import de.blinkt.openvpn.R
import de.blinkt.openvpn.fragments.OpenVpnPreferencesFragment
import de.blinkt.openvpn.VpnProfile

class Settings_IP : OpenVpnPreferencesFragment(), Preference.OnPreferenceChangeListener {
    private lateinit var mIPv4: EditTextPreference
    private lateinit var mIPv6: EditTextPreference
    private lateinit var mUsePull: SwitchPreference
    private lateinit var mOverrideDNS: CheckBoxPreference
    private lateinit var mSearchdomain: EditTextPreference
    private lateinit var mDNS1: EditTextPreference
    private lateinit var mDNS2: EditTextPreference
    private lateinit var mNobind: CheckBoxPreference
    override fun onResume() {
        super.onResume()


        // Make sure default values are applied.  In a real app, you would
        // want this in a shared function that is used to retrieve the
        // SharedPreferences wherever they are needed.
        PreferenceManager.setDefaultValues(
            requireActivity(),
            R.xml.vpn_ipsettings, false
        )

        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.vpn_ipsettings)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        /* Bind the preferences early to avoid loadingSetting which is called
         * from the superclass to access an uninitialised earlyinit property
         */
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onBindPreferences() {
        super.onBindPreferences()
        bindPreferences()
        loadSettings()
    }

    private fun bindPreferences() {
        mIPv4 = findPreference("ipv4_address")!!
        mIPv6 = findPreference("ipv6_address")!!
        mUsePull = findPreference("usePull")!!
        mOverrideDNS = findPreference("overrideDNS")!!
        mSearchdomain = findPreference("searchdomain")!!
        mDNS1 = findPreference("dns1")!!
        mDNS2 = findPreference("dns2")!!
        mDNS1.summaryProvider = DNSSummaryProvider()
        mDNS2.summaryProvider = DNSSummaryProvider()
        mNobind = findPreference("nobind")!!
        mUsePull.onPreferenceChangeListener = this
        mOverrideDNS.onPreferenceChangeListener = this
    }


    override fun loadSettings() {
        // Since we maybe not have preferences bound yet, check if we actually have them bound.
        if (!this::mUsePull.isInitialized) {
            return;
        }
        if (mProfile.mAuthenticationType == VpnProfile.TYPE_STATICKEYS) mUsePull.isEnabled =
            false else mUsePull.isChecked = mProfile.mUsePull
        mIPv4.text = mProfile.mIPv4Address
        mIPv6.text = mProfile.mIPv6Address
        mDNS1.text = mProfile.mDNS1
        mDNS2.text = mProfile.mDNS2
        mOverrideDNS.isChecked = mProfile.mOverrideDNS
        mSearchdomain.text = mProfile.mSearchDomain
        mNobind.isChecked = mProfile.mNobind


        //mUsePull.setEnabled(mProfile.mAuthenticationType != VpnProfile.TYPE_STATICKEYS);
        mUsePull.isEnabled = true
        if (mProfile.mAuthenticationType == VpnProfile.TYPE_STATICKEYS) mUsePull.isChecked = false

        // Sets Summary
        setDNSState()
    }

    override fun saveSettings() {
        mProfile.mUsePull = mUsePull.isChecked
        mProfile.mIPv4Address = mIPv4.text
        mProfile.mIPv6Address = mIPv6.text
        mProfile.mDNS1 = mDNS1.text
        mProfile.mDNS2 = mDNS2.text
        mProfile.mOverrideDNS = mOverrideDNS.isChecked
        mProfile.mSearchDomain = mSearchdomain.text
        mProfile.mNobind = mNobind.isChecked
    }

    override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
        if (preference === mUsePull || preference === mOverrideDNS)
            if (preference === mOverrideDNS) {
                // Set so the function gets the right value
                mOverrideDNS.isChecked = (newValue as Boolean)
            }
        setDNSState()
        saveSettings()
        return true
    }

    override fun onPause() {
        super.onPause()
        saveSettings()
    }

    private fun setDNSState() {
        val enabled: Boolean
        mOverrideDNS.isEnabled = mUsePull.isChecked
        enabled = if (!mUsePull.isChecked) true else mOverrideDNS.isChecked
        mDNS1.isEnabled = enabled
        mDNS2.isEnabled = enabled
        mSearchdomain.isEnabled = enabled
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {

    }
}

class DNSSummaryProvider : Preference.SummaryProvider<Preference> {
    override fun provideSummary(preference: Preference): CharSequence {
        val ep = preference as EditTextPreference
        var summary = ep.text ?: ""
        if (summary == "8.8.4.4" || summary == "8.8.8.8" || summary == "2001:4860:4860::8888" || summary == "2001:4860:4860::8844" )
            summary += " (dns.google.com)"
        else if (summary.startsWith("2606:4700:4700::") || summary.startsWith("1.1.1.") || summary.startsWith("1.0.0."))
            summary += " (Cloudflare)"

        else if (summary.startsWith("9.9.9.") || summary.startsWith("2620:fe::"))
            summary += " (Quad9)"
        else if (summary.isEmpty())
            summary = "(not set)"

        return summary
    }

}
